PAPER.md

```diff
@@ -0,0 +1,378 @@
++                                                               +
+|                                                               |
++                                                               +
+|                                                               |
++                                                               +
+|                                                               |
++                                                               +
+|                                                               |
++                                                               +
+|                           source(256)                         |
++       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|       |                                                       |
++-+-+-+-+                                                       +
+|                                                               |
++                                                               +
+|                                                               |
++                                                               +
+|                                                               |
++                                                               +
+|                                                               |
++                                                               +
+|                                                               |
++                                                               +
+|                                                               |
++                                                               +
+|                        destination(256)                       |
++       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|       |         headers len(16)       |                       |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+                       +
+|                             headers                           |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                          data len(32)                         |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                          data                         |       |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|    signature len(16)  |                                       |
++-+-+-+-+-+-+-+-+-+-+-+-+                                       +
+|                            signature                          |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+```
+
+* `type` field for this message will be the unsigned integer `4`.
+
+* `uuid` field its an [UUID](https://www.rfc-editor.org/rfc/rfc4122.html) v4 that needs to be added to every message by the application, so previous messages can be easy referenced and found later.
+
+* `prev hash` field will be the SHA256 sum of the last acknowledged message in the conversation.
```

Comment:
I guess this field will be 0 in the `genesis` message of each conversation. I think it is worth identifying this case if at some point we dedice to store the conversation somewhere (blockchain), because we will need to retrieve the blocks based on some criteria.

Reply:
Its a great point. Fixed at af609c3eb3ee1ed1799c741b6a13aea0be765bc7

Now, thinking a bit more in your words (need your point of view) ... This could be happening among 2 separate peers without internet, so i think we are only able to maintain a continued linking of messages inside each conversation. That means treating each conversation as an independent fragment of a message chain. 2 ideas regarding to this:

* Would be good to have an "conversation UUID" on each message ? So easier and faster to group messages by other systems.

* What you mentioned brings more ideas about the kind of _future_ "message chain data structure" we could use. I am thinking that _maybe_, we could organise the separate conversations as completely independent "small chains" which are running in parallel. Then we could glue them together in a time index. We know the conversations are initiated by an zeroed `prev hash` field, then just need to find all the messages with that conversation `UUID`. Then compose the chain based on `prev hash` again. The first message of the conversation (timestamp field) would indicate its position in the commented time index.

Sorry about bringing this "futuristic" ideas here, but would like to ensure the viability of a future message chain. Let me know your thoughts !

Reply:
Added the fields at 66aaebb5a4b87385d221cda623706b2cc83051d8
Added the the blockchain idea at b4df167f071c76c63fe4b1105a1916debee4ba43

This is BTW just an initial idea. Lets start building, experimenting and adapting.